### PR TITLE
[routing] Akka.Routing refactoring

### PR DESCRIPTION
**Pull request by sean-gilliam**
This PR refactors the classes in the Akka.Routing namespace to take
advantage of newer C# features.  It also simplifies some of the logic in
the various routers.

### CODE REVIEW

File: src/core/Akka/Routing/Broadcast.cs [hunk 1]
```diff
@@ -153,38 +153,28 @@ public override RouterConfig WithFallback(RouterConfig routerConfig)
 
         private RouterConfig OverrideUnsetConfig(RouterConfig other)
         {
-            if (other is NoRouter)
+            if (other is Pool pool)
```
Comment (sean-gilliam) [inline src/core/Akka/Routing/Broadcast.cs]: everything that isn't a ```Pool``` returned ```this``` , so this is the only real check we need. Same with the other routers.

File: src/core/Akka/Routing/ConsistentHashRouter.cs [hunk 2]
```diff
@@ -185,42 +185,43 @@ public override Routee Select(object message, Routee[] routees)
...
-                    _log.Value.Warning("Couldn't route message with consistent hash key [{0}] due to [{1}]", hashData,
-                        ex.Message);
+                    _log.Value.Warning("Couldn't route message with consistent hash key [{0}] due to [{1}]", hashData, ex.Message);
                     return Routee.NoRoutee;
                 }
-            };
+            }
 
-            if (_hashMapping(message) != null)
+            var hashMapping = _hashMapping(message);
```
Comment (sean-gilliam) [inline src/core/Akka/Routing/ConsistentHashRouter.cs]: This delegate was being called twice which should return the same value each time. So I stored it in a variable so it only needed to be called once.

Comment (Aaronontheweb) [inline src/core/Akka/Routing/ConsistentHashRouter.cs]: Nice catch

**Comment by Aaronontheweb** [comment]
looks great @sean-gilliam - love seeing the C#7 features cutting down on code bloat

File: src/core/Akka/Routing/ConsistentHash.cs [hunk 3]
```diff
@@ -289,35 +289,39 @@ public ISurrogated FromSurrogate(ActorSystem system)
...
-                return BitConverter.GetBytes((double)obj);
-            if (obj is decimal)
-                return new BitArray(decimal.GetBits((decimal)obj)).ToBytes();
-            if (obj is Guid)
-                return ((Guid)obj).ToByteArray();
-            return obj;
+            switch (obj)
+            {
+                case null:
+                    return new byte[] { 0 };
```
Comment (maxcherednik) [inline src/core/Akka/Routing/ConsistentHash.cs]: I don't know the context, and how often this one called and maybe it's super micro, but seems like a constant?

**Pull request merged**